### ISSUE TITLE
False negative TS2739 on generic object key constraint for variable declaration vs. assignment

### ISSUE THREAD

**Issue by warent** [microsoft/TypeScript#39983]
**TypeScript Version:**  3.9.7, 4.0.0-beta

**Search Terms:**
`is:issue is:open label:Bug  generic keys `

**TL;DR**
[Playground Link](https://www.typescriptlang.org/play?ts=4.0.0-beta#code/C4TwDgpgBA9gRgKygXigbwFBSgawiARgC4oA7AVwFs4IAnAbi13wCYSLq7Hs8QBmdlRoMMAX0YYAxjFIBnYFADmEUnQCGwGLRRQAPABUoEAB7AVAE1lR5tAJalFAGigBVHWigBtANJR7UfQBdQU5tUQA+AApeWRJ9T0CAShI3ZHCmGIA6WghzckkISKZsSOkAGzKISWBnXmd7cxNElHCoSMxsTqhMnvLK6sdizs9eYL9SRuNBrtFE6c60USZmtSsXCWk5BWVVWg1cnR31TVpIzwAiADMYeDVac6TGTfkoOFoYPFJ9cAgAYQALKo4EjwJCoI57E5nK43OB3B6JJ4yF4Ady0OHsim+kABQJBiEOKmOuXoQA)

**Code**

```ts
type obj = {
  key1: number;
  key2: number;
  key3: number;
};

const generator = <T extends string, U = { [K in T]: number }>(keys: T[]): U =>
  keys.reduce(
    (collect, key, index) => ({
      ...collect,
      [key]: index,
    }),
    {}
  ) as U;

const generated = generator(["key1", "key2", "key3"]);
const myObject: obj = generated;
```

This presents no type error. No bug, this is expected behavior. Now change the lines:

```ts
const generated = generator(["foobar"]);
const myObject: obj = generated;
```

And we get the following error.

```
demo.ts:18:7 - error TS2739: Type '{ foobar: number; }' is missing the following properties from type 'obj': key1, key2, key3

18 const myObject: obj = generated;
         ~~~~~~~~


Found 1 error.
```

This is still expected behavior, no bug here. However, things break when we make a slight adjustment and assign the variable directly:

```ts
const myObject: obj = generator(["foobar"]);
```

This removes the type error even though it is expected to be firing exactly the same as above.

**Comment by DanielRosenwasser** [comment]
`U` doesn't have anything to infer from on the input position, so TypeScript will infer from its contextual type (`obj`). Instead of a default, you could constrain the type based on `T`.

```ts
type obj = {
  key1: number;
  key2: number;
  key3: number;
};

const generator = <T extends string, U extends { [K in T]: number }>(keys: T[]): U =>
  keys.reduce(
    (collect, key, index) => ({
      ...collect,
      [key]: index,
    }),
    {}
  ) as U;

const generated = generator(["foobar"]);
const brokenTypeCheck: obj = generator(["foobar"]);
const workingTypeCheck: obj = generated;
const myObject: obj = generator(["foobar"]);
```

Even better, don't use generic type parameters that don't occur in the input positions; it's typically a sign that you're relying on cast-like behavior.

```ts
type obj = {
  key1: number;
  key2: number;
  key3: number;
};

const generator = <T extends string>(keys: T[]): { [K in T]: number } =>
  keys.reduce(
    (collect, key, index) => ({
      ...collect,
      [key]: index,
    }),
    {}
  ) as any;

const generated = generator(["foobar"]);
const brokenTypeCheck: obj = generator(["foobar"]);
const workingTypeCheck: obj = generated;
const myObject: obj = generator(["foobar"]);
```

**Comment by warent** [comment]
Very strange. I don't understand why `U extends` is behaving so differently from `U =` and why the latter is inferring type from the variable it's being assigned to. Guess I'm getting to the point where it's time to sit down and really read through the docs.

Thanks for your help!

**Comment by DanielRosenwasser** [comment]
Well `=` is just a default - it doesn't imply anything about the constraint. That is a little counter-intuitive, but it makes some more code expressible.
